llvm/lib/Analysis/InlineAdvisor.cpp

```diff
@@ -63,6 +65,65 @@ static cl::opt<bool>
 
 extern cl::opt<InlinerFunctionImportStatsOpts> InlinerFunctionImportStats;
 
+/// Flag to path of dynamic libarary
+class DLInlineAdvisorInfo {
+  typedef std::unique_ptr<InlineAdvisor> (*DLInlineAdivsorFactory_t)(
+      Module &M, FunctionAnalysisManager &FAM, LLVMContext &Context,
+      std::unique_ptr<InlineAdvisor> OriginalAdvisor, InlineContext IC,
+      std::string &DLCTX);
+
+  cl::opt<std::string> *Path;
+  void *Handle;
+  DLInlineAdivsorFactory_t Factory;
+
+public:
+  DLInlineAdvisorInfo() : Handle(nullptr), Factory(nullptr) {}
+  DLInlineAdvisorInfo(cl::opt<std::string> *Path)
+      : Path(Path), Handle(nullptr), Factory(nullptr) {}
+  std::unique_ptr<InlineAdvisor> factory(Module &M, FunctionAnalysisManager &FAM, LLVMContext &Context,
+          std::unique_ptr<InlineAdvisor> OriginalAdvisor, InlineContext IC,
+          std::string &DLCTX) {
+    if (!Handle && Path->getNumOccurrences() > 0) {
+      Handle = dlopen(Path->c_str(), RTLD_LAZY);
+      if (!Handle) {
+        errs() << "Cannot open library: " << dlerror() << '\n';
+      } else {
+        LLVM_DEBUG(dbgs() << "Library opened successfully\n";)
+        Factory =
+            (DLInlineAdivsorFactory_t)dlsym(Handle, "DLInlineAdvisorFactory");
+        if (!Factory) {
+          errs() << "Cannot find function DLInlineAdvisorFactory: " << dlerror()
+                 << "\n";
+        } else {
+          LLVM_DEBUG(dbgs() << "Function DLInlineAdvisorFactory found successfully\n";)
+        }
+      }
+    }
+    if (Factory) {
+      return Factory(M, FAM, Context, std::move(OriginalAdvisor), IC, DLCTX);
+    }
+  }
+
+  ~DLInlineAdvisorInfo() {
+    if (Handle)
+      dlclose(Handle);
+  }
+};
+
+static DLInlineAdvisorInfo DLInlineAdvisor(&DLInlineAdvisorPath);
+
+static cl::opt<std::string>
+    DLInlineAdvisorPath("dl-inline-advisor-path",
+                         cl::desc("Path to a dynamic library that can be used "
+                                  "instead of the default inline advisor."),
+                         cl::value_desc("dynamic library path."));
+
+static cl::opt<std::string> DLInlineAdvisorCTX(
+    "dl-inline-advisor-ctx", cl::init(""),
+    cl::desc(
+        "String passed to dynamic inline advisor to modify it's behaviour."),
```

Comment:
how is this supposed to be used? Are these flags for the DLAdvisor

Reply:
So the idea was that you can do "-dl-inline-advisor-ctx <some string here>" and that get's passed to the dlinlineadvisor constructor as the final parameter, that way you could have some way of modifying the behaviour from the command line